concepts/null-undefined/about.md

```diff
@@ -0,0 +1,152 @@
+
+`undefined` appears in different contexts.
+
+- If a variable is declared, it is `undefined` initially.
+- If you try to access a value for a non-existing key in an object, you get `undefined`.
+- If a function does not return a value, the result is `undefined`.
+
+```javascript
+let name;
+console.log(name);
+// => undefined
+
+let obj = { greeting: 'hello world' };
+console.log(obj.missingKey);
+// => undefined
+
+function returnNothing() {
+  return;
+}
+console.log(returnNothing());
+// => undefined
+```
+
+You can check whether a variable is undefined using the strict equality operator `===` or the `typeof` operator.
+
+```javascript
+let name;
+
+name === undefined;
+// => true
+
+typeof name === 'undefined';
+// => true
+```
+
+It is not recommended to manually assign `undefined` to a variable, always use `null` instead to make it clear you set an empty value intentionally.
+
+## Optional Chaining
+
+As mentioned above, accessing a non-existent key in an object returns `undefined` in JavaScript.
+However if you try to retrieve a nested value and the parent key does not exist, the evaluation of the nested key is performed on `undefined` and leads to `TypeError: Cannot read property ... of undefined`.
+Theoretically, you would always need to check the parent key exists before you can try to retrieve the nested key.
+This was often done with expressions like `obj.key && obj.key.nestedKey`.
+Now imagine how this looks for deeply nested values.
+
+To solve this problem, [optional chaining][mdn-optional-chaining] was added to the language specification in 2020.
+With the optional chaining operator `?.` you can ensure that JavaScript only tries to access the nested key if the parent was not `null` or `undefined`.
+Otherwise `undefined` is returned.
+
+```javascript
+const obj = {
+  address: {
+    street: 'Trincomalee Highway',
+    city: 'Batticaloa',
+  },
+};
+
+obj.residence;
+// => undefined
+
+obj.address.zipCode;
+// => undefined
+
+obj.residence.street;
+// => TypeError: Cannot read property 'street' of undefined
+
+obj.residence?.street;
+// => undefined
+
+obj.residence?.street?.number;
+// => undefined
+```
+
+## Nullish Coalescing
+
+There are situations where you want to apply a default value in case a variable is null or undefined.
+In the past this was often times done utilizing lazy evaluation of the OR operator `||`.
+This has the disadvantage that the default value is applied in all cases where the variable is [falsy][mdn-falsy] (e.g. `''` or `0`), not only when it is null or undefined.
+This can easily cause unexpected outcomes.
+
+```javascript
+let amount = null;
+amount = amount || 1;
+// => 1
+
+amount = 0;
+amount = amount || 1;
+// => 1
+```
+
+To address this, the [nullish coalescing operator][mdn-nullish-coalescing] `??` was introduced.
+Just like optional chaining, it was added to the language specification in 2020.
```

Comment:
I think we should add a note saying older versions of nodejs or browsers may not support this.

Reply:
I discussed this with @SleeplessByte and we agreed that instead of adding this note in a lot of concepts over and over again we want to add a section about this issue in general to one of the track wide documentation files. Then we can also explain things like "caniuse", babel, proposal stages etc there that help the student to navigate this issue. Once we have that we could of course link to it from different places.